agent/rpcclient/health/health.go

```diff
@@ -68,5 +70,74 @@ func (c *Client) getServiceNodes(
 	if !ok {
 		panic("wrong response type for cachetype.HealthServicesName")
 	}
-	return *value, md, nil
+
+	return filterTags(filterNodeMeta(value, req), req), md, nil
+}
+
+func filterTags(out *structs.IndexedCheckServiceNodes, req structs.ServiceSpecificRequest) structs.IndexedCheckServiceNodes {
+	if len(req.ServiceTags) == 0 || len(out.Nodes) == 0 {
+		return *out
+	}
+	tags := make([]string, 0, len(req.ServiceTags))
+	for _, r := range req.ServiceTags {
+		// DNS has the bad habit to setting [""] for ServiceTags
+		if r != "" {
+			tags = append(tags, strings.ToLower(r))
+		}
```

Comment:
Does this logic exist in the current server side implementation? I had a quick look but couldn't find it. Maybe we should fix this in `dns.go` instead to not send the empty tag?

It looks like the server code uses the `TagFilter` field to determine if the tags should be used or not. It seems strange to me to have separate a separate boolean field enable the filtering instead of checking the length of the slice of tags, but I guess that's how it works.

To keep this consistent I guess this will need to both check if `TagFilter` is true, and also include `req.ServiceTag` (although that's been deprecated for a while, so maybe we don't need to support it here).

Reply:
@dnephin Ok, DONE, I also fixed in DNS by sending `[]` instead of `[""]` when no tags are set in requests